File: client/src/lib/google-sheets.ts
import { google } from "googleapis";

function configuredSpreadsheetId() {
  return process.env.GOOGLE_SHEETS_ID ?? "";
}

function oauthCreds() {
  return {
    clientId: process.env.GOOGLE_OAUTH_CLIENT_ID ?? "",
    clientSecret: process.env.GOOGLE_OAUTH_CLIENT_SECRET ?? "",
    redirectUri:
      process.env.GOOGLE_OAUTH_REDIRECT_URI ??
      "http://localhost:5001/oauth2callback",
    refreshToken: process.env.GOOGLE_OAUTH_REFRESH_TOKEN ?? "",
  };
}

const PREFERRED_TAB = "Cases";
const HEADER_ROW = [
  "Case Number",
  "Status",
  "Service Type",
  "Service Code",
  "Patient ID",
  "Insurer Org ID",
  "Created At",
  "Updated At",
  "Patient Name",
  "Disease",
  "Claim Purpose",
  "Drive URL",
];
const LAST_COL = "L";

export type CaseSheetRow = {
  case_number: string;
  status: string;
  service_type: string;
  service_code?: string | null;
  patient_id: string;
  insurer_org_id: string;
  created_at?: string | null;
  updated_at?: string | null;
  patient_name?: string | null;
  disease?: string | null;
  claim_purpose?: string | null;
  drive_url?: string | null;
};

function quotedTab(name: string) {
  return `'${name.replace(/'/g, "''")}'`;
}

function sheetsClient() {
  const { clientId, clientSecret, redirectUri, refreshToken } = oauthCreds();
  if (!clientId || !clientSecret || !refreshToken) {
    throw new Error(
      "Google OAuth is not configured. Set GOOGLE_OAUTH_CLIENT_ID, GOOGLE_OAUTH_CLIENT_SECRET, and GOOGLE_OAUTH_REFRESH_TOKEN.",
    );
  }
  const auth = new google.auth.OAuth2(clientId, clientSecret, redirectUri);
  auth.setCredentials({ refresh_token: refreshToken });
  return google.sheets({ version: "v4", auth });
}

function toValues(row: CaseSheetRow) {
  return [
    row.case_number,
    row.status,
    row.service_type,
    row.service_code ?? "",
    row.patient_id,
    row.insurer_org_id,
    row.created_at ?? "",
    row.updated_at ?? "",
    row.patient_name ?? "",
    row.disease ?? "",
    row.claim_purpose ?? "",
    row.drive_url ?? "",
  ];
}

function googleMessage(err: unknown) {
  if (err && typeof err === "object") {
    const maybe = err as {
      message?: string;
      response?: { data?: { error?: { message?: string; status?: string } } };
    };
    const api = maybe.response?.data?.error;
    if (api?.message) {
      return `${api.status ?? "Sheets"}: ${api.message}`;
    }
    if (maybe.message) return maybe.message;
  }
  return "Google Sheets sync failed";
}

async function ensureSheet() {
  const spreadsheetId = configuredSpreadsheetId();
  if (!spreadsheetId) {
    throw new Error("GOOGLE_SHEETS_ID is missing from client/.env.local.");
  }

  const sheets = sheetsClient();
  const spreadsheet = await sheets.spreadsheets.get({
    spreadsheetId,
    fields: "sheets(properties(title))",
  });
  const titles = (spreadsheet.data.sheets ?? [])
    .map((sheet) => sheet.properties?.title)
    .filter((title): title is string => Boolean(title));
  const sheetName = titles.includes(PREFERRED_TAB)
    ? PREFERRED_TAB
    : titles[0] ?? PREFERRED_TAB;

  if (!titles.includes(sheetName)) {
    await sheets.spreadsheets.batchUpdate({
      spreadsheetId,
      requestBody: {
        requests: [{ addSheet: { properties: { title: sheetName } } }],
      },
    });
  }

  const header = await sheets.spreadsheets.values.get({
    spreadsheetId,
    range: `${quotedTab(sheetName)}!A1:${LAST_COL}1`,
  });
  const firstRow = header.data.values?.[0] ?? [];
  const empty = firstRow.every((cell) => !String(cell ?? "").trim());
  if (empty) {
    await sheets.spreadsheets.values.update({
      spreadsheetId,
      range: `${quotedTab(sheetName)}!A1:${LAST_COL}1`,
      valueInputOption: "USER_ENTERED",
      requestBody: { values: [HEADER_ROW] },
    });
  }

  return { sheets, spreadsheetId, sheetName };
}

export async function appendCaseRow(row: CaseSheetRow) {
  try {
    const { sheets, spreadsheetId, sheetName } = await ensureSheet();
    const result = await sheets.spreadsheets.values.append({
      spreadsheetId,
      range: `${quotedTab(sheetName)}!A:${LAST_COL}`,
      valueInputOption: "USER_ENTERED",
      insertDataOption: "INSERT_ROWS",
      requestBody: { values: [toValues(row)] },
    });
    if (!result.data.updates?.updatedRows) {
      throw new Error("Google Sheets accepted the request but wrote 0 rows.");
    }
  } catch (err) {
    throw new Error(googleMessage(err));
  }
}

File: ai-server/app/services/google_sheets.py
from typing import Any, Dict, List, Optional
from urllib.parse import quote

import httpx

from app.services.env import clean_env, is_dry_run

PREFERRED_TAB = "Cases"
HEADER_ROW = [
    "Case Number",
    "Status",
    "Service Type",
    "Service Code",
    "Patient ID",
    "Insurer Org ID",
    "Created At",
    "Updated At",
    "Patient Name",
    "Disease",
    "Claim Purpose",
    "Drive URL",
]
LAST_COL = "L"
TOKEN_URL = "https://oauth2.googleapis.com/token"
SHEETS_BASE = "https://sheets.googleapis.com/v4/spreadsheets"


def _spreadsheet_id() -> str:
    return clean_env("GOOGLE_SHEETS_ID")


def _quoted(name: str) -> str:
    return quote(f"'{name.replace(chr(39), chr(39) * 2)}'", safe="")


def _access_token() -> str:
    creds = {
        "client_id": clean_env("GOOGLE_OAUTH_CLIENT_ID"),
        "client_secret": clean_env("GOOGLE_OAUTH_CLIENT_SECRET"),
        "refresh_token": clean_env("GOOGLE_OAUTH_REFRESH_TOKEN"),
    }
    if not all(creds.values()):
        raise RuntimeError(
            "Google OAuth is not configured. Set GOOGLE_OAUTH_CLIENT_ID, "
            "GOOGLE_OAUTH_CLIENT_SECRET, and GOOGLE_OAUTH_REFRESH_TOKEN."
        )
    response = httpx.post(
        TOKEN_URL,
        data={**creds, "grant_type": "refresh_token"},
        timeout=20.0,
    )
    response.raise_for_status()
    token = response.json().get("access_token")
    if not token:
        raise RuntimeError("Google OAuth did not return an access token")
    return token


def _values(row: Dict[str, Any]) -> List[str]:
    return [
        str(row.get("case_number") or ""),
        str(row.get("status") or ""),
        str(row.get("service_type") or ""),
        str(row.get("service_code") or ""),
        str(row.get("patient_id") or row.get("payer_id") or ""),
        str(row.get("insurer_org_id") or ""),
        str(row.get("created_at") or ""),
        str(row.get("updated_at") or ""),
        str(row.get("patient_name") or ""),
        str(row.get("disease") or row.get("payer_id") or ""),
        str(row.get("claim_purpose") or row.get("denial_reason") or ""),
        str(row.get("drive_url") or ""),
    ]


def _resolve_sheet(client: httpx.Client, spreadsheet_id: str) -> str:
    meta = client.get(
        f"{SHEETS_BASE}/{spreadsheet_id}",
        params={"fields": "sheets(properties(title))"},
    )
    meta.raise_for_status()
    titles = [
        sheet.get("properties", {}).get("title")
        for sheet in meta.json().get("sheets", [])
        if sheet.get("properties", {}).get("title")
    ]
    sheet_name = PREFERRED_TAB if PREFERRED_TAB in titles else (titles[0] if titles else PREFERRED_TAB)
    header = client.get(
        f"{SHEETS_BASE}/{spreadsheet_id}/values/{_quoted(sheet_name)}!A1:{LAST_COL}1"
    )
    header.raise_for_status()
    first = (header.json().get("values") or [[]])[0]
    if not any(str(cell).strip() for cell in first):
        update = client.put(
            f"{SHEETS_BASE}/{spreadsheet_id}/values/{_quoted(sheet_name)}!A1:{LAST_COL}1",
            params={"valueInputOption": "USER_ENTERED"},
            json={"values": [HEADER_ROW]},
        )
        update.raise_for_status()
    return sheet_name


def upsert_case_row(row: Dict[str, Any]) -> Optional[str]:
    if is_dry_run():
        return None
    spreadsheet_id = _spreadsheet_id()
    if not spreadsheet_id:
        return "GOOGLE_SHEETS_ID is missing"

    try:
        with httpx.Client(
            headers={"Authorization": f"Bearer {_access_token()}"},
            timeout=20.0,
        ) as client:
            sheet_name = _resolve_sheet(client, spreadsheet_id)
            existing = client.get(
                f"{SHEETS_BASE}/{spreadsheet_id}/values/{_quoted(sheet_name)}!A:A"
            )
            existing.raise_for_status()
            numbers = [cells[0] if cells else "" for cells in existing.json().get("values") or []]
            case_number = str(row.get("case_number") or "")
            values = {"values": [_values(row)]}
            try:
                row_index = numbers.index(case_number)
            except ValueError:
                row_index = -1

            if row_index <= 0:
                append = client.post(
                    f"{SHEETS_BASE}/{spreadsheet_id}/values/{_quoted(sheet_name)}!A:{LAST_COL}:append",
                    params={
                        "valueInputOption": "USER_ENTERED",
                        "insertDataOption": "INSERT_ROWS",
                    },
                    json=values,
                )
                append.raise_for_status()
                return None

            update = client.put(
                f"{SHEETS_BASE}/{spreadsheet_id}/values/{_quoted(sheet_name)}!A{row_index + 1}:{LAST_COL}{row_index + 1}",
                params={"valueInputOption": "USER_ENTERED"},
                json=values,
            )
            update.raise_for_status()
            return None
    except Exception as err:
        return str(err)
